EpiAware/src/EpiInfModels/ODEProcess.jl

```diff
@@ -0,0 +1,253 @@
+```
+Where `S` is the proportion of the population that is susceptible, `E` is the proportion of the
+population that is exposed, `I` is the proportion of the population that is infected and `R` is
+the proportion of the population that is recovered. The parameters are the infectiousness `β`,
+the incubation rate `α` and the recovery rate `γ`.
+
+```julia
+using EpiAware, OrdinaryDiffEq, Distributions
+
+# Create an instance of SIRParams
+seirparams = SEIRParams(
+    tspan = (0.0, 30.0),
+    infectiousness = LogNormal(log(0.3), 0.05),
+    incubation_rate = LogNormal(log(0.1), 0.05),
+    recovery_rate = LogNormal(log(0.1), 0.05),
+    initial_prop_infected = Beta(1, 99)
+)
+```
+
+# Usage example with `ODEProcess` and predefined SIR model
+
+In this example we define an `ODEProcess` object using the predefined `SIRParams` model from
+above. We then generate latent infections using the `generate_latent_infs` function, and refit
+the model using a `Turing` model.
+
+We assume that the latent infections are observed with a Poisson likelihood around their
+ODE model prediction. The population size is `N = 1000`, which we put into the `sol2infs` function,
+which maps the ODE solution to the number of infections. Recall that the `EpiAware` default SIR
+implementation assumes the model is in density/proportion form. Also, note that since the `sol2infs`
+function is a link function that maps the ODE solution to the expected number of infections we also
+apply the `LogExpFunctions.logaddexp` function to ensure that the expected number of infections is non-negative.
+The `logaddexp` function generalises the `softplus` function, `logaddexp(1, x) == softplus(x)`.
+
+First, we define the `ODEProcess` object which combines the SIR model with the `sol2infs` link
+function and the solver options.
+
+```julia
+using Turing, LogExpFunctions
+N = 1000.0
+
+sir_process = ODEProcess(
+    params = sirparams,
+    sol2infs = sol -> N .* logaddexp.(0.001, sol[2, :]),
+    solver_options = Dict(:verbose => false, :saveat => 1.0)
+)
+```
+
+Next, we create a `Turing` model for the infection generating process using the `generate_latent_infs` function.
+This function remakes the ODE problem with the provided initial conditions and parameters, solves it using the
+specified solver, and then transforms the solution into latent infections using the `sol2infs`
+function.
+
+NB: The `nothing` argument is a dummy latent process, e.g. a log-Rt time series, that is not
+used in the SIR model, but might be used in other models.
+
+```julia
+sir_infections_mdl = generate_latent_infs(sir_process, nothing)
+```
+
+We can generate some test data from the model by sampling from the model and then calling the model
+
+```julia
+# Sampled parameters
+θ = rand(sir_infections_mdl)
+expected_infections = (sir_infections_mdl | θ)()
+test_data = expected_infections .|> λ -> rand(Poisson(λ))
+```
+
+Now, we can refit the model using a `PoissonError` observation model
+
+```julia
```

Comment:
I think we need to move away from using julia and only use @example

Reply:
This is a really good point (and see below where this would have caught an outdated approach.

However, I think that `@example` doesn't mean execution inside docstrings.

Options:

1. Replace with `jldoctest` and end a block with something like `nothing` and just checking nothing. This make the doctest a test against the code not failing.
2. Have actual targets for doctest
3. Reduce size of docstrings and put complicated code into a (pluto or documenter-flavoured md) tutorial.

Reply:
I'll offer option 1. to concretise this a bit; but happy to vary.

Reply:
Implemented option 1

Reply:
Not resolved this conversation because interested in @seabbs view here.